### PR TITLE
ASTs keep track of their root node

### LINKED ISSUE #39: Edit script generation fails if input files have different parsers
This was discovered in PR #36 but reproducing does not require the `-g` switch, so the bug must have been in the code base for some time.

Example input files:

```
$ cat tests/add.calc
1+2

$ cat tests/Hello.txt
public class Hello {
    public static void main(String[] args) {
        System.out.println("Hello, world!");
    }
}
```

Example run:

```
$ diffract -a -m Myers tests/add.calc tests/Hello.txt 

INFO:diffract: Selecting the Myers matching algorithm.
^~
 ~
 Expr
  Term
   Factor
    INT 1
    ~
  PLUS +
  ~
  Expr
   Term
    Factor
     INT 2
     ~
      WHITESPACE 

      ~

document
 paragraph
  WORD public
  SPACE  
  paragraph
   WORD class
   SPACE  
   paragraph
    WORD Hello
    SPACE  
    paragraph
     WORD {
 END 

 document
  paragraph
   SPACE     
   paragraph
    WORD public
    SPACE  
    paragraph
     WORD static
     SPACE  
     paragraph
      WORD void
      SPACE  
      paragraph
       WORD main(String[]
       SPACE  
       paragraph
        WORD args)
        SPACE  
        paragraph
         WORD {
  END 

  document
   paragraph
    SPACE         
    paragraph
     WORD System.out.println("Hello,
     SPACE  
     paragraph
      WORD world!");
   END 

   document
    paragraph
     SPACE     
     paragraph
      WORD }
    END 

    document
     paragraph
      WORD }
     END 


thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::panicking::panic
             at /checkout/src/libcore/panicking.rs:51
  10: <core::option::Option<T>>::unwrap
             at /checkout/src/libcore/macros.rs:20
  11: <diffract::matchers::MappingStore<T>>::generate_edit_script
             at ./src/lib/matchers.rs:345
  12: diffract::main
             at src/main.rs:374
  13: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  14: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  15: main
  16: __libc_start_main
```

### LINKED ISSUE #41: Relax the expectation that Node 0 in an AST is the root
At the moment several places in the code base assume that Node 0 is the root node of the AST. In fact, `INS`ert operations can insert a new root node during the edit script generation phase of the differencer. 

The `INS` operation needs a specific test case for replacing the root, the graphviz generator, etc. need careful thought.

### PR DESCRIPTION
**Pull request by snim2**
**NEEDS SQUASHING**

Fixes #39
Fixes #41

### CODE REVIEW

File: src/lib/action.rs [hunk 1]
```diff
@@ -174,7 +174,10 @@ impl RenderJson for Insert {
         json.push(" ".repeat(indent) + "{");
         json.push(format!("{}\"action\": \"insert\",", ind_s));
         json.push(format!("{}\"tree\": {},", &ind_s, self.node.id()));
-        json.push(format!("{}\"parent\": {},", &ind_s, self.new_parent.id()));
+        match self.new_parent {
```
Comment (ltratt) [inline src/lib/action.rs]: Would this be simpler as something like:

```rust
json.push(format!("{}\"parent\": {},", &ind_s, self.new_parent.map_or(|x| x.id(), "")));
```

?

File: src/lib/matchers.rs [hunk 2]
```diff
@@ -335,13 +336,26 @@ impl<T: Clone + Debug + Eq + 'static> MappingStore<T> {
...
-            let y = self.to_arena[x].parent().unwrap(); // x is not a root.
             let mut w = root; // Overwritten later.
             // Insertion phase.
-            if !new_mappings.to.contains_key(&x) {
+            if !new_mappings.to.contains_key(&x) && x.is_root(&self.to_arena) {
+                let k = self.find_pos(x, &new_mappings, &to_in_order);
+                debug!("Edit script: INS {} {:?} No parent",
+                       self.to_arena[x].label,
+                       self.to_arena[x].value);
+                w = self.from_arena.new_node(self.to_arena[x].value.clone(),
```
Comment (ltratt) [inline src/lib/matchers.rs]: What exactly are we cloning here? [I'm not saying it's wrong, but I'm not really sure if we need to clone, or whether a reference might do.]

File: src/lib/sequence.rs [hunk 3]
```diff
@@ -158,9 +158,20 @@ mod tests {
             assert!(lcss(&vec![], &store.from_arena, &vec![], &store.to_arena, &eq).is_empty());
             return;
         }
-        let root = NodeId::new(0);
-        let base = root.children(&store.from_arena).collect::<Vec<NodeId>>();
-        let diff = root.children(&store.to_arena).collect::<Vec<NodeId>>();
+        assert!(!store.from_arena.is_empty());
+        assert!(!store.to_arena.is_empty());
+        let base = store
```
Comment (ltratt) [inline src/lib/sequence.rs]: Weird indentation of the following lines.

File: src/lib/matchers.rs (hunk 2 above)
Comment (snim2) [reply src/lib/matchers.rs]: `value` is of a type `T`, it's dependent on the grammar...

Comment (ltratt) [reply src/lib/matchers.rs]: OK.

File: src/lib/action.rs (hunk 1 above)
Comment (snim2) [reply src/lib/action.rs]: Interesting. This doesn't compile, I think because you are saying "either give me a usize, or a str"...

Comment (ltratt) [reply src/lib/action.rs]: I did say "something like" ;) You can convert the usize to a string easily enough.

**Comment by snim2** [comment]
OK, ready to squash?

**Comment by snim2** [comment]
If you do that then one branch gets a `&str` and the other gets a `String` . If you do `""` and `p.id().to_string().as_str()` then the borrow checker complains about a borrowed value not living long enough.

**Comment by snim2** [comment]
Last commit: better or worse?

**Comment by ltratt** [comment]
Worse: now we're heap allocating stuff when we don't need to :(

**Comment by snim2** [comment]
OK, good to squash?

**Comment by ltratt** [comment]
Yep!

**Comment by snim2** [comment]
Squashed.

**Pull request merged**